napari/utils/naming.py

```diff
@@ -61,16 +61,31 @@ def magic_name(value, *, path_prefix):
     name : str or None
         Name of the variable, if found.
     """
-    frame = inspect.currentframe().f_back
-    code = frame.f_code
-
-    while code.co_filename.startswith(path_prefix):
-        frame = frame.f_back
-        code = frame.f_code
-
-    varmap = ChainMap(frame.f_locals, frame.f_globals)
-    names = *code.co_varnames, *code.co_names
-
-    for name in names:
-        if name.isidentifier() and name in varmap and varmap[name] is value:
-            return name
+    try:
+        frame = inspect.currentframe()
+        if inspect.isframe(frame):
+            frame = frame.f_back
+
+        while (
+            inspect.isframe(frame)
+            and inspect.iscode(frame.f_code)
+            and frame.f_code.co_filename.startswith(path_prefix)
+        ):
+            frame = frame.f_back
+
+        if inspect.isframe(frame) and inspect.iscode(frame.f_code):
+            varmap = ChainMap(frame.f_locals, frame.f_globals)
+            names = *frame.f_code.co_varnames, *frame.f_code.co_names
+
+            for name in names:
+                if (
+                    name.isidentifier()
+                    and name in varmap
+                    and varmap[name] is value
+                ):
+                    return name
+    finally:
+        # Prevent stack frame leak
+        del frame
```

Comment:
Why this is need? Why GC is not enough? After go out from function reference counter will go down. 
I suggest rather `frame=None` to not break debugger.

Reply:
See here:

https://docs.python.org/3/library/inspect.html#the-interpreter-stack
https://bugs.python.org/issue543148

Reply:
If I good understand You should move bellow `return None` in try clause. Otherwise you could meet some problem with execution code when frame was deleted. 

I also suggest use `frame.clear()` protected by `None` check. Reason is mentioned in documentation:

> If you want to keep the frame around (for example to print a traceback later), you can also break reference cycles by using the frame.clear() method.

As you see issue which you point is for Python in version 2.1. Later one gc clear this, but with some delay.

Reply:
I moved the `return None` into the try block. `frame.clear()` causes a RuntimeError though.

Reply:
I see. It looks like there should be created bug report to python documentation.

Reply:
The problem is that frames being referenced are still executing, so we cannot clear them.

Reply:
I run test with simple code 
```python 
import inspect
import os

print(os.getpid())

def leak():
    frame = inspect.currentframe()

while 1:
    leak()
```
and record memory usage using `procpath`  on python 3.8. 

It looks like there is no memory leak: 
![Zrzut ekranu z 2020-10-08 12-25-10](https://user-images.githubusercontent.com/3826210/95446999-6374aa80-0961-11eb-8a92-21d6a393fdc2.png)

Its also look like `del` only remove frame reference. I could access to variable values with debugger even when frame was deleted in `finally` step. 

Fro my test it looks like `del frame` is obsolete but do not have negative impact.

Reply:
Maybe try a deeper call stack? Are there any downsides to having `try` / `finally` / `del frame` ?

Reply:
Only which I see is code complexity. I will try later

Reply:
I see that you're using RSS to check, but I think we also might want to use something that can probe the Python interpreter and garbage collection more directly.

I'm not sure if Napari will run on anything other than CPython. However, the presence of a cycle detector is an interpreter specific matter.

Given that this pattern is currently recommended in the docs for inspect, I would lean towards keeping it. It would be worthwhile to inquire upstream whether that note still applies though.

Reply:
@jni I'm curious what your thoughts are here?

Reply:
I'm afraid I'm out of my stack depth. 😂 It would be good at least to add comments above the `del frame` line, linking to this discussion as well as the references therein so that future code archeologists can understand the current design. Based on the CPython docs I agree with @mkitti that the cycle detector is an implementation detail. We probably depend on CPython vs other interpreters like, a lot 😂 but we should try to avoid hardcoding that dependency as much as possible.

I'll note one difference between the current code and the code in the linked CPython docs: `frame = inspect.currentframe()` happens outside the `try:` in the docs. Should we mimic that?

Reply:
I have added comments with references and moved the initial line out of try block. Thanks for taking a look at this again, @jni.